plugins/inputs/t128_tank/reader.go

```diff
@@ -165,30 +160,93 @@ func (lc *LineContext) Accumulated() string {
 	return lc.accumulating.line
 }
 
-func yieldValidLines(topic string, indexFile *IndexFile) []string {
-	readNext := indexFile.last()
-	if !contains(specialIndices, readNext) {
-		readNext++
-	}
-	runner := processRunner(topic, readNext)
-	if !runner.started {
-		panic(fmt.Sprintf("could not start tank-cli: %s", runner.result))
-	}
-	var lines []string
-	for _, line := range inde(indexFile, runner.watchStdout()) {
-		lines = append(lines, strings.TrimSpace(line))
-	}
-	if !runner.result.isSuccess {
-		panic(fmt.Sprintf("tank-cli stopped: %s", runner.result.message))
-	}
-	return lines
+func (bf BoundaryFault) Error() string {
+	return "Boundary fault"
+}
+
+func YieldValidLines(topic int, indexFile *IndexFile, isMainThread bool, tankPath string, shouldExpectEveryLine bool) <-chan string {
+	result := make(chan string)
+
+	go func() {
+		readNext := indexFile.Last()
+
+		if !contains(specialIndices, readNext) {
+			readNext = readNext + 1
+		}
+
+		cliCommand := buildCLICommand(tankPath, topic, readNext)
+		cmd := exec.Command(cliCommand[0], cliCommand[1:]...)
+		cmd.SysProcAttr = &syscall.SysProcAttr{Setpgid: true}
+		stdout, err := cmd.StdoutPipe()
+		if err != nil {
+			fmt.Errorf("Error creating stdout pipe: " + err.Error())
+			close(result)
+			return
+		}
+
+		if err := cmd.Start(); err != nil {
+			fmt.Errorf("Error starting tank-cli: " + err.Error())
+			close(result)
+			return
+		}
+
+		defer cmd.Wait()
+
+		var wg sync.WaitGroup
+		wg.Add(1)
+		go func() {
```

Comment:
I don't think three levels of go routines is necessary here. If you copied it, that's fine. Let's not change it right now. But I think the same functionality can be achieved by just deleting this second level routine. When the for loop below completes, the work is done and the function can exit. There's no need for the wait group etc. to accomplish it (unless I'm missing something). Again, if this is directly copied, let's just go with what's here for now.